lib/phoenix/controller.ex

```diff
@@ -433,22 +433,72 @@ defmodule Phoenix.Controller do
 
   Raises `Plug.Conn.AlreadySentError` if `conn` is already sent.
   """
-  @spec put_view(Plug.Conn.t, atom) :: Plug.Conn.t
+  @spec put_view(Plug.Conn.t, [{atom | String.t(), atom}] | atom) :: Plug.Conn.t
+  def put_view(%Plug.Conn{state: state} = conn, formats) when state in @unsent and is_list(formats) do
+    put_private_formats(conn, :phoenix_view, :replace, formats)
+  end
   def put_view(%Plug.Conn{state: state} = conn, module) when state in @unsent do
-    put_private(conn, :phoenix_view, module)
+    put_private_formats(conn, :phoenix_view, :replace, _: module)
   end
 
   def put_view(%Plug.Conn{}, _module), do: raise AlreadySentError
 
+  defp put_private_formats(conn, priv_key, kind, formats) when kind in [:new, :replace] do
+    formats =
+      Enum.into(formats, %{}, fn
+        {:_, layout} -> {:_, layout}
+        {format, layout} -> {to_string(format), layout}
+      end)
+
+    update_in(conn.private, fn private ->
+      existing = private[priv_key] || %{}
+      new_formats =
+        case kind do
+          :new -> Map.merge(formats, existing)
+          :replace -> Map.merge(existing, formats)
+        end
+
+      Map.put(private, priv_key, new_formats)
+    end)
+  end
+
+  defp get_private_format(conn, priv_key) do
+    format = conn.private[:phoenix_format]
+
+    case conn.private[priv_key] do
+      %{:_ => value} ->
+        value
+
+      %{^format => value} ->
+        value
+
+      value when value in [nil, false] ->
+        false
+
+      %{} = formats when map_size(formats) == 1 ->
+        {_format, value} = (Enum.at(formats, 0))
+        value
```

Comment:
I think no longer necessary if `:_`is always first?

```suggestion
```

Reply:
we still need this check to avoid the `get_format` lookup if there is only a single one